ibm/resource_ibm_network_gateway.go

```diff
@@ -0,0 +1,849 @@
+	public_vlan_id := d.Get("public_vlan_id").(int)
+
+	if public_vlan_id > 0 {
+		order.Hardware[0].PrimaryNetworkComponent = &datatypes.Network_Component{
+			NetworkVlan: &datatypes.Network_Vlan{Id: sl.Int(public_vlan_id)},
+		}
+	}
+
+	private_vlan_id := d.Get("private_vlan_id").(int)
+	if private_vlan_id > 0 {
+		order.Hardware[0].PrimaryBackendNetworkComponent = &datatypes.Network_Component{
+			NetworkVlan: &datatypes.Network_Vlan{Id: sl.Int(private_vlan_id)},
+		}
+	}
+
+	return order, nil
+}
+
+func waitForNoGatewayActiveTransactions(id int, meta interface{}) (interface{}, error) {
+	log.Printf("Waiting for Gateway (%d) to have zero active transactions", id)
+	service := services.GetHardwareServerService(meta.(ClientSession).SoftLayerSession())
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"retry", "active"},
+		Target:  []string{"idle"},
+		Refresh: func() (interface{}, string, error) {
+			bm, err := service.Id(id).Mask("id,activeTransactionCount").GetObject()
+			if err != nil {
+				return false, "retry", nil
+			}
+
+			if bm.ActiveTransactionCount != nil && *bm.ActiveTransactionCount == 0 {
+				return bm, "idle", nil
+			}
+			return bm, "active", nil
+
+		},
+		Timeout:        24 * time.Hour,
+		Delay:          10 * time.Second,
+		MinTimeout:     1 * time.Minute,
+		NotFoundChecks: 24 * 60,
+	}
+
+	return stateConf.WaitForState()
+}
+
+// Network gateways or Bare metal creation does not return a  object with an Id.
+// Have to wait on provision date to become available on server that matches
+// hostname and domain.
+// http://sldn.softlayer.com/blog/bpotter/ordering-bare-metal-servers-using-softlayer-api
+func waitForNetworkGatewayProvision(d *datatypes.Hardware, meta interface{}) (interface{}, error) {
+	hostname := *d.Hostname
+	domain := *d.Domain
+	log.Printf("Waiting for Gateway (%s.%s) to be provisioned", hostname, domain)
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"retry", "pending"},
+		Target:  []string{"provisioned"},
+		Refresh: func() (interface{}, string, error) {
+			service := services.GetAccountService(meta.(ClientSession).SoftLayerSession())
+			bms, err := service.Filter(
+				filter.Build(
+					filter.Path("hardware.hostname").Eq(hostname),
+					filter.Path("hardware.domain").Eq(domain),
+				),
+			).Mask("id,provisionDate").GetHardware()
+			if err != nil {
+				return false, "retry", nil
+			}
+
+			if len(bms) == 0 || bms[0].ProvisionDate == nil {
+				return datatypes.Hardware{}, "pending", nil
+			} else {
+				return bms[0], "provisioned", nil
+			}
+		},
+		Timeout:        24 * time.Hour,
+		Delay:          10 * time.Second,
+		MinTimeout:     1 * time.Minute,
+		NotFoundChecks: 24 * 60,
+	}
+
+	return stateConf.WaitForState()
+}
+
+func flattenVLANInstances(list []datatypes.Network_Gateway_Vlan) []map[string]interface{} {
```

Comment:
This was not need, because the flattenVLANInstances function was intended to read the VLANS already configured in the Network Gateway, but that information is not relevant in this case. I removed it from the code